<a href="https://colab.research.google.com/github/thfruchart/1nsi/blob/main/19/COURS_ENTIERS_Binaire_Hexa.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

L'information traitée par un ordinateur est de nature **binaire** : soit le courant passe, soit il ne passe pas!

Ces deux états peuvent être représentés par les nombres **1** et **0**.

En informatique, les données numériques sont représentées avec des suites de chiffres binaires (*binary digit* ou *bit*).

Dans ce chapitre, on s'intéresse à la représentation des **nombres entiers** en binaire (puis en hexadécimal). 

#### Activité PAPIER : écrire des nombres avec plus ou moins de symboles (10, 2, ou 16) !

In [2]:
print (' décimal |   binaire |  hexadécimal')
for i in range(300):
    print("{0:>8d} |{0:>10b} | {0:>3X}".format(i))

 décimal |   binaire |  hexadécimal
       0 |         0 |   0
       1 |         1 |   1
       2 |        10 |   2
       3 |        11 |   3
       4 |       100 |   4
       5 |       101 |   5
       6 |       110 |   6
       7 |       111 |   7
       8 |      1000 |   8
       9 |      1001 |   9
      10 |      1010 |   A
      11 |      1011 |   B
      12 |      1100 |   C
      13 |      1101 |   D
      14 |      1110 |   E
      15 |      1111 |   F
      16 |     10000 |  10
      17 |     10001 |  11
      18 |     10010 |  12
      19 |     10011 |  13
      20 |     10100 |  14
      21 |     10101 |  15
      22 |     10110 |  16
      23 |     10111 |  17
      24 |     11000 |  18
      25 |     11001 |  19
      26 |     11010 |  1A
      27 |     11011 |  1B
      28 |     11100 |  1C
      29 |     11101 |  1D
      30 |     11110 |  1E
      31 |     11111 |  1F
      32 |    100000 |  20
      33 |    100001 |  21
      34 |    100010 |  22
      35 |    10001

# Nombres écrits en binaire (ou en base 2)

Pour écrire un nombre en binaire, on n'utilise uniquement les deux chiffres 0 et 1.

En Python, le préfixe `0b`... sert à écrire un nombre en binaire. 

En mathématiques, on notera plutôt $(...)_2$.

Par exemple : le nombre entier 7 écrit en binaire, sera noté :
*  $(111)_2$ en mathématiques
* `0b111` en Python

Pour comprendre comment fonctionne cette écriture des nombres, observer l'exécution des cellules suivantes  :

In [3]:
0b10

2

In [4]:
0b100

4

In [5]:
0b1000

8

In [6]:
0b10000

16

#### Question


Comment calculer `0b10000000000`  soit $(10000000000)_2$ c'est à dire 1 suivi de dix 0?

In [7]:
0b10000000000

1024

#### quelques calculs

Pour chaque cellule : calculer d'abord le résultat de l'opération, puis vérifier en exécutant la cellule. 

In [ ]:
0b1000 + 0b10

In [ ]:
0b10000 + 0b100

In [ ]:
0b100 +0b10 +0b1

**Application**: sur PAPIER

En décomposant sous forme d'une somme de nombre de la forme 10...0, calculer:

* $(101010)_2$
* $(1111)_2$

## Conversion vers la base 10 : décimal

**Principe** : chaque chiffre d'un nombre binaire représente une puissance de 2

On peut repérer le "poids" de chaque chiffre, en lui faisant correspondre une puissance de 2:

|RANG|$4$|$3$|$2$|$1$|$0$|
|:--|:--:|:--:|:--:|:--:|:--:|
|chiffres binaires|1|1|0|1|0|

|POIDS|$2^4$|$2^3$|$2^2$|$2^1$|$2^0$|
|:--|:--:|:--:|:--:|:--:|:--:|
|chiffres binaires|1|1|0|1|0|

ce qui donne : 

|poids|16|8|4|2|1|
|:--|:--:|:--:|:--:|:--:|:--:|
|chiffres binaires|1|1|0|1|0|

donc : $(11010)_2 =  16+ 8+ 2  = 26$ 

In [ ]:
0b11010

## Conversion depuis la base 10

On cherche à décomposer un nombre N en somme de puissance de 2
* écrire les puissances de 2 jusqu'à atteindre (au moins la moitié de) N
* par soustractions successives, écrire N comme somme de puissances de 2.


Exemple : 183 peut se décomposer en somme de puissances de 2


|poids|128|64|32|16|8|4|2|1|
|:--|:--:|:--:|:--:|:--:|:--:|:--:|:--:|:--:|
|chiffres binaires|1|0|1|1|0|1|1|1|

$183 = 128+32+16+4+2+1 = (10110111)_2$

In [ ]:
(183-128, 55-32, 23-16, 7-4, 3-2)

(55, 23, 7, 3, 1)

Avec Python, on peut vérifier à l'aide de la fonction `bin`

In [ ]:
bin(183)

'0b10110111'

# Nombres écrits en hexadécimal (base 16)

En Python, le préfixe `0x`... sert à écrire un nombre au format hexadécimal.

En mathématiques, on utilise la notation $(...)_{16}$

La notation hexadécimale repose sur le fait que chaque chiffre représente une puissance de 16.

Observer les exemples suivants : 

In [8]:
0x10

16

In [9]:
0x100

256

In [10]:
16**2

256

In [11]:
0x1000

4096

In [12]:
16**3

4096

In [13]:
0x10000 == 16**4

True

## Conversion vers la base 10 : hexadécimal =>  décimal

**Principe** : chaque chiffre d'un nombre écrit en hexadécimal représente une puissance de 16.

On peut repérer le "poids" de chaque chiffre, en lui faisant correspondre la puissance de 16 qui correspond à son rang :

|poids|$16^4$|$16^3$|$16^2$|$16^1$|$16^0$|calcul|écriture décimale
|:--|:--:|:--:|:--:|:--:|:--:|:----------:|:--|
|chiffres hexadécimaux|||1|1|1|$16^2+16^1+16^0$|273|

In [ ]:
0x111

273

**ATTENTION** 

Pour écrire les nombres en hexadécimal (base 16), on utilise **16 chiffres** : 
* les chiffres habituels de 0 à 9
* les "chiffres" de A à F :
   * A pour dix
   * B pour onze
   * C pour douze
   * D pour treize
   * E pour quatorze
   * F pour quinze

On peut repérer le "poids" de chaque chiffre, en lui faisant correspondre une puissance de 16:

|poids|$16^4$|$16^3$|$16^2$|$16^1$|$16^0$|calcul|écriture décimale
|:--|:--:|:--:|:--:|:--:|:--:|:----------:|:--|
|chiffres hexadécimaux|1|0|A|8|F|65536+0+2560+128+15 |68239|
|calculs|1x$16^4$|0|10x$16^2$|8x$16^1$|15x$16^0$| ||

* $(10A8F)_{16}=68 239$

In [ ]:
0x10A8F

## Conversion vers la base 2 : hexadécimal => binaire

**Propriété**

Puisque $16 = 2^4$, **un** chiffre en base 16 correspond à **quatre** chiffres en base 2 (un "quartet"). 

| décimal|hexadécimal| binaire|
|:---------:|:------:|:------:|
|0|0|0000|
|1|1|0001|
|2|2|0010|
|3|3|0011|
|4|4|0100|
|5|5|0101|
|6|6|0110|
|7|7|0111|
|8|8|1000|
|9|9|1001|
|10|A|1010|
|11|B|1011|
|12|C|1100|
|13|D|1101|
|14|E|1110|
|15|F|1111|

Pour convertir un nombre entier naturel depuis la base 16 (hexadécimal) vers la base 2 (binaire), il suffit de convertir chaque chiffre hexadécimal en quatre chiffres binaires.

Exemple : $(F01)_{16} = (1111 \, 0000 \, 0001)_2$

In [ ]:
0xF01 == 0b111100000001

## Conversion depuis la base 2 : binaire => hexadécimal

Le même principe permet de convertir directement un nombre écrit en binaire, vers la base 16 (hexadécimal) : 
il s'agit de grouper les chiffres binaires par quatre **à partir du chiffre des unités**.

Exemple 1:

$(1010101)_2 = (0101 \, 0101)_2 =(55)_{16}$